### PR TITLE
Added hook to verify password to allow custom password check

### PR DESCRIPTION
**Pull request by Peterede**
I have added a hook to allow overriding the password check in order to provide a different password evaluation.  I am using this so that each user can have a different password.

### CODE REVIEW

File: lib/Room.php [hunk 1]
```diff
@@ -633,6 +633,15 @@ public function changeInCall($sessionId, $active) {
 	 * @return bool
 	 */
 	public function verifyPassword($password) {
+		$event = new GenericEvent(null, [
+			'password' => $password,
+			'hasPassword' => $this->hasPassword(),
+		]);
```
Comment (fancycode) [inline lib/Room.php]: You should pass the room as subject of the event, so the event handler knows which room is checking the password and has access to all parameters:
```php
new GenericEvent($this, [
    'password' => $password,
]);
```

**Comment by Peterede** [comment]
I am not sure on how to create the test, should it be something like:

		$this->dispatcher->expects($this->at(0))
			->method('dispatch')
			->with('\OCA\Spreed\Room::verifyPassword');

File: lib/Room.php [hunk 2]
```diff
@@ -595,6 +604,13 @@ public function joinRoomGuest($password, $passedPasswordProtection = false) {
 		return $sessionId;
 	}
 
+	/**
+	 * @param int $length
+	 */
+	public function generateSessionId($length) {
```
[the remark this answers is not in the document]
Comment (Peterede) [reply lib/Room.php]: This is used in the hook preJoinRoomGuest to generate a sessionId and return it from the event.  secureRandom is private in the Room class so this function is needed for this purpose.  This is part of my work on allowing guests to be assigned moderator status so in preJoinRoomGuest I create a sessionId and update the database for the user based upon email address and return the session.

[the remark this answers is not in the document]
Comment (Peterede) [reply lib/Room.php]: I will remove it for this PR

**Comment by Peterede** [comment]
I have added tests for verifying the password and testing the dispatch of the event

File: lib/Room.php [hunk 3]
```diff
@@ -633,7 +634,20 @@ public function changeInCall($sessionId, $active) {
...
 	 */
 	public function verifyPassword($password) {
-		return !$this->hasPassword() || $this->hasher->verify($password, $this->password);
+		$event = new GenericEvent($this, [
+			'password' => $password
+		]);
+		$this->dispatcher->dispatch(self::class . '::verifyPassword', $event);
+		if ($event->hasArgument('result')) {
+			$result = $event->getArgument('result');
+			return $result;
```
Comment (nickvergessen) [inline lib/Room.php]: I would prefer:
```php
return [
	'result' => $result['result'] ?? false,
	'url' => $result['url'] ?? ''
];
```
here, just to make sure that the values we require in our code are returned

File: lib/Controller/PageController.php [hunk 4]
```diff
@@ -147,10 +147,22 @@ public function index($token = '', $callUser = '', $password = '') {
 				}
 
 				if ($requirePassword) {
-					if ($password !== '' && $room->verifyPassword($password)) {
+					$passwordVerification = [
```
Comment (nickvergessen) [inline lib/Controller/PageController.php]: Useless assignment

File: lib/Controller/PageController.php [hunk 5]
```diff
@@ -147,10 +147,22 @@ public function index($token = '', $callUser = '', $password = '') {
...
+						'url' => ''
+					];
+					
+					$passwordVerification = $room->verifyPassword($password);
+
+					if ($passwordVerification['result']) {
 						$this->session->setPasswordForRoom($token, $token);
 					} else {
-						return new TemplateResponse($this->appName, 'authenticate', [], 'guest');
+						if ($passwordVerification['url'] == '') {
```
Comment (nickvergessen) [inline lib/Controller/PageController.php]: `empty($passwordVerification['url'])` or `=== ''`

File: appinfo/info.xml [hunk 6]
```diff
@@ -44,7 +44,7 @@ And in the works for the [coming versions](https://github.com/nextcloud/spreed/m
 	<screenshot>https://raw.githubusercontent.com/nextcloud/spreed/master/docs/contacts-menu.png</screenshot>
 
 	<dependencies>
-		<nextcloud min-version="14" max-version="14" />
+		<nextcloud min-version="13" max-version="14" />
```
Comment (nickvergessen) [inline appinfo/info.xml]: Revert this please

File: lib/Controller/PageController.php [hunk 7]
```diff
@@ -194,10 +202,22 @@ protected function guestEnterRoom($token, $password) {
 
 		$this->session->removePasswordForRoom($token);
 		if ($room->hasPassword()) {
-			if ($password !== '' && $room->verifyPassword($password)) {
+			$passwordVerification = [
```
Comment (nickvergessen) [inline lib/Controller/PageController.php]: useless assignment

File: lib/Controller/PageController.php [hunk 8]
```diff
@@ -194,10 +202,22 @@ protected function guestEnterRoom($token, $password) {
...
+				'url' => ''
+			];
+
+			$passwordVerification = $room->verifyPassword($password);
+
+			if ($passwordVerification['result']) {
 				$this->session->setPasswordForRoom($token, $token);
 			} else {
-				return new TemplateResponse($this->appName, 'authenticate', [], 'guest');
+				if ($passwordVerification['url'] == '') {
```
Comment (nickvergessen) [inline lib/Controller/PageController.php]: `===`

File: tests/php/PasswordVerificationTest.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+use OCP\Security\IHasher;
+use OCP\Security\ISecureRandom;
+use Test\TestCase;
+
+class PasswordVerificationTest extends TestCase {
+
+	public function testVerifyPassword() {
+		$dispatcher = \OC::$server->getEventDispatcher();
+		
+		$dispatcher->addListener('OCA\Spreed\Room::verifyPassword', function(GenericEvent $event) {
```
Comment (nickvergessen) [inline tests/php/PasswordVerificationTest.php]: missing import of `Symfony\Component\EventDispatcher\GenericEvent`

1) OCA\Spreed\Tests\php\Signaling\BackendNotifierTest::testRoomInCallChanged
TypeError: Argument 1 passed to OCA\Spreed\Tests\php\PasswordVerificationTest::OCA\Spreed\Tests\php\{closure}() must be an instance of OCA\Spreed\Tests\php\GenericEvent, instance of Symfony\Component\EventDispatcher\GenericEvent given, called in /drone/src/github.com/nextcloud/server/3rdparty/symfony/event-dispatcher/EventDispatcher.php on line 212

/drone/src/github.com/nextcloud/server/apps/spreed/tests/php/PasswordVerificationTest.php:35
/drone/src/github.com/nextcloud/server/3rdparty/symfony/event-dispatcher/EventDispatcher.php:212
/drone/src/github.com/nextcloud/server/3rdparty/symfony/event-dispatcher/EventDispatcher.php:44
/drone/src/github.com/nextcloud/server/apps/spreed/lib/Room.php:645
/drone/src/github.com/nextcloud/server/apps/spreed/lib/Room.php:583
/drone/src/github.com/nextcloud/server/apps/spreed/tests/php/Signaling/BackendNotifierTest.php:287

File: tests/php/PasswordVerificationTest.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+namespace OCA\Spreed\Tests\php;
+
+use OCA\Spreed\Config;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IConfig;
+use OCP\Security\IHasher;
+use OCP\Security\ISecureRandom;
+use Test\TestCase;
+
+class PasswordVerificationTest extends TestCase {
```
Comment (nickvergessen) [inline tests/php/PasswordVerificationTest.php]: missing database group:
```php
/**
 * @group DB
 */
```

**Comment by nickvergessen** [comment]
And #1077

**Pull request merged**